Sources/NIO/ControlMessage.swift

```diff
@@ -172,18 +173,57 @@ struct ControlMessageParser {
     }
     
     mutating func receiveMessage(_ controlMessage: UnsafeControlMessage) {
-        if controlMessage.level == IPPROTO_IP && controlMessage.type == ControlMessageParser.ipv4TosType {
+        if controlMessage.level == IPPROTO_IP {
+            receiveIPv4Message(controlMessage)
+        } else if controlMessage.level == IPPROTO_IPV6 {
+            receiveIPv6Message(controlMessage)
+        }
+    }
+
+    mutating func receiveIPv4Message(_ controlMessage: UnsafeControlMessage) {
+        if controlMessage.type == ControlMessageParser.ipv4TosType {
             if let data = controlMessage.data {
                 assert(data.count == 1)
                 precondition(data.count >= 1)
                 let readValue = CInt(data[0])
                 self.ecnValue = .init(receivedValue: readValue)
             }
-        } else if controlMessage.level == IPPROTO_IPV6 && controlMessage.type == IPV6_TCLASS {
+        } else if controlMessage.type == Posix.IP_PKTINFO {
+            if let data = controlMessage.data {
+                guard let info = data.bindMemory(to: in_pktinfo.self).baseAddress else {
+                    return
+                }
+
+                var addr = sockaddr_in()
+                addr.sin_family = sa_family_t(NIOBSDSocket.AddressFamily.inet.rawValue)
+                addr.sin_port = in_port_t(0).bigEndian
+                addr.sin_addr = info.pointee.ipi_addr
+                self.packetInfo = NIOPacketInfo(destinationAddress: SocketAddress(addr, host: ""), interfaceIndex: Int(info.pointee.ipi_ifindex))
+            }
+
+        }
+    }
+
+    mutating func receiveIPv6Message(_ controlMessage: UnsafeControlMessage) {
+        if controlMessage.type == IPV6_TCLASS {
             if let data = controlMessage.data {
                 let readValue = ControlMessageParser._readCInt(data: data)
                 self.ecnValue = .init(receivedValue: readValue)
             }
+        } else if controlMessage.type == Posix.IPV6_PKTINFO {
+            if let data = controlMessage.data {
+                guard let info = data.bindMemory(to: in6_pktinfo.self).baseAddress else {
```

Comment:
This presumably doesn't work on Linux, as we need to use the `CNIOLinux` variant of this structure, right?

Reply:
Can we do `load` here as well?

Reply:
It does work on Linux. I added a `typealias` to the `CNIOLinux` variant in `System.swift`: https://github.com/apple/swift-nio/pull/1888/files#diff-3f800fb8f45dc37abc935011731dc02fe9bb2eb245b7f943be8a8214af930bc5R29 I'm open to other ways to solve this but this struct is behind a macro that appears to have the same issue as `_GNU_SOURCE`: https://bugs.swift.org/browse/SR-6772

Reply:
Great, that's what I missed. The `typealias` is fine.